source/MRMesh/MRPointsSave.cpp

```diff
@@ -124,74 +131,130 @@ tl::expected<void, std::string> toCtm( const PointCloud& points, std::ostream& o
-        return s.good() ? size : 0;
-    }, &out );
+        std::function<bool( float )> callbackFn{};
+        std::ostream* stream;
+        size_t sum{ 0 };
+        size_t blockSize{ 0 };
+        size_t maxSize{ 0 };
+    } saveData;
+    if ( callback )
+    {
+        saveData.callbackFn = [callback, &saveData] ( float progress )
+        {
+            // calculate full progress in partical-linear scale (we don't know compressed size and it less than real size)
+            progress = ( saveData.sum + progress * saveData.blockSize ) / saveData.maxSize;
+            float newProgress = 0.f;
+            for ( int i = 0; i < 5; ++i )
+            {
+                if ( progress < 0.2f )
+                {
+                    newProgress += progress / 0.2f * 0.7f * ( 1 - newProgress );
+                    break;
+                }
+                else
+                {
+                    progress = ( progress - 0.2f ) / 0.8f;
+                    newProgress += ( 1 - newProgress ) * 0.7f;
+                }
+            }
+            return callback( newProgress );
+        };
+    }
+    saveData.stream = &out;
+    saveData.maxSize = points.points.size() * sizeof( Vector3f );
+    ctmSaveCustom( context, [] ( const void* buf, CTMuint size, void* data )
+    {
+        SaveData& saveData = *reinterpret_cast< SaveData* >( data );
+        std::ostream& outStream = *saveData.stream;
+        saveData.blockSize = size;
+
+        const bool cancel = !MR::writeByBlocks( outStream, (const char*) buf, size, saveData.callbackFn, 1u << 12 );
+        saveData.sum += size;
+        if ( cancel )
+            return 0u;
+
+        return outStream.good() ? size : 0;
+    }, &saveData );
 
     if ( !out || ctmGetError( context ) != CTM_NONE )
         return tl::make_unexpected( std::string( "Error saving in CTM-format" ) );
 
+    if ( callback )
+        callback( 1.f );
     return {};
-
 }
 
-tl::expected<void, std::string> toPts( const PointCloud& points, const std::filesystem::path& file )
+tl::expected<void, std::string> toPts( const PointCloud& points, const std::filesystem::path& file, ProgressCallback callback )
 {
     std::ofstream out( file, std::ofstream::binary );
     if ( !out )
         return tl::make_unexpected( std::string( "Cannot open file for writing " ) + utf8string( file ) );
 
-    return toPts( points, out );
+    return toPts( points, out, callback );
 }
 
-tl::expected<void, std::string> toPts( const PointCloud& points, std::ostream& out )
+tl::expected<void, std::string> toPts( const PointCloud& points, std::ostream& out, ProgressCallback callback )
 {
     out << "BEGIN_Polyline\n";
+    const float pointsNum = float( points.validPoints.count() );
+    int pointIndex = 0;
     for ( auto v : points.validPoints )
+    {
         out << points.points[v] << "\n";
+        ++pointIndex;
+        if ( !( pointIndex % 1000 ) && callback && !callback( float( pointIndex ) / pointsNum ) )
```

Comment:
better
```
if ( callback && !( pointIndex % 1000 ) && !callback( float( pointIndex ) / pointsNum ) )
```
to avoid division when no callback, and in other similar places

Reply:
fixed